Fix Apiman Issue #732

jhauray opened the pull request:
This pull request include the following changes :
- VertxPluginRegistry : Init "pluginsDir" and "pluginRepositories" with conf file values.
- VertxPluginRegistry : Configure http client following protocol (http or https), and use system properties to configure proxy if defined.
- DefaultPluginRegistry : As see with @EricWittmann , update createTempPluginsDir with code from VertxPluginRegistry .

A new Unit test class is added : VertxPluginRegistryTest.

All Vert.X conf files are updated with commented "pluginsDir" and "pluginRepositories" parameters.

jhauray commented:
@EricWittmann,

Snyk seems to fail on gateway/platforms/vertx3/vertx3/pom.xml check.

The same in previous Pull requests.

Do you know why ?

Snyk FAQ report to recreate the pull request : https://support.snyk.io/hc/en-us/articles/360000920678-Failed-to-detect-issues

What should I do ?

msavy commented:
You can ignore it for now. I'll need to look to bump the vert.x version
separately.

With the new security integrations in GH it might be worth turning off Snyk
now anyway

On Mon, 6 Jan 2020, 08:36 Jérémy HAURAY, <notifications@github.com> wrote:

> @EricWittmann <https://github.com/EricWittmann>,
>
> Snyk seems to fail on gateway/platforms/vertx3/vertx3/pom.xml check.
>
> The same in previous Pull requests.
>
> Do you know why ?
>
> Snyk FAQ report to recreate the pull request :
> https://support.snyk.io/hc/en-us/articles/360000920678-Failed-to-detect-issues
>
> What should I do ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apiman/apiman/pull/788?email_source=notifications&email_token=AADHMWPHLC6BZXFQDYFF5ODQ4LUO7A5CNFSM4J574XX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIEZDMQ#issuecomment-571052466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADHMWJNCQQ62EL2FNMYEKTQ4LUO7ANCNFSM4J574XXQ>
> .
>

jhauray commented:
> You can ignore it for now. I'll need to look to bump the vert.x version separately. With the new security integrations in GH it might be worth turning off Snyk now anyway
> […](#)
> On Mon, 6 Jan 2020, 08:36 Jérémy HAURAY, ***@***.***> wrote: @EricWittmann <https://github.com/EricWittmann>, Snyk seems to fail on gateway/platforms/vertx3/vertx3/pom.xml check. The same in previous Pull requests. Do you know why ? Snyk FAQ report to recreate the pull request : https://support.snyk.io/hc/en-us/articles/360000920678-Failed-to-detect-issues What should I do ? — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#788?email_source=notifications&email_token=AADHMWPHLC6BZXFQDYFF5ODQ4LUO7A5CNFSM4J574XX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIEZDMQ#issuecomment-571052466>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADHMWJNCQQ62EL2FNMYEKTQ4LUO7ANCNFSM4J574XXQ> .

OK,

Thank you for the reply @msavy ! So, I let you reviewing the pull request.

And best wishes 🎊 !!

msavy commented on gateway/engine/core/src/main/java/io/apiman/gateway/engine/impl/DefaultPluginRegistry.java:
```diff
@@ -104,11 +105,7 @@ public DefaultPluginRegistry(File pluginsDir, Set<URI> pluginRepositories) {
         try {
-            @SuppressWarnings("nls")
-            File tempDir = File.createTempFile("_apiman", "plugins");
-            tempDir.delete();
```
I seem to recall this was some sort of hack for Windows xx @EricWittmann ?

EricWittmann replied:
I don't recall that.

EricWittmann replied:
I think this change is OK.

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -62,8 +57,34 @@ private static File getTempPluginsDir() {
+
+        super(config);
+
+        this.vertx=vertx;
```
Style nit: please use spaces on assignment

jhauray replied:
done!

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -62,8 +57,34 @@ private static File getTempPluginsDir() {
+        this.vertx=vertx;
+
+        //Get HTTPS Proxy settings (useful for local dev tests and corporate CI)
+        String sslProxyHost = System.getProperty("https.proxyHost", "none");
```
Interesting idea!

jhauray replied:
And useful when internet outbound is protected by a corporate proxy.

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -72,30 +93,73 @@ public VertxPluginRegistry(Vertx vertx, VertxEngineConfig vxEngineConfig, Map<St
+                                      final IAsyncResultHandler<File> handler) {
+
+        HttpClient client;
+        HttpClientOptions clientOpts=new HttpClientOptions();
```
Style nit: spaces on equals assignment

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -72,30 +93,73 @@ public VertxPluginRegistry(Vertx vertx, VertxEngineConfig vxEngineConfig, Map<St
+            if (port == -1) port = 443;
+
+            //If artifact host is localhost
+            if(artifactUrl.getHost().equals("localhost") ||artifactUrl.getHost().equals("127.0.0.1")){
```
Style nit: spaces on if statement, including `||`

Generally spaces around all these constructs (I'll stop pointing each one out, I'm sure you get the idea!).

msavy replied:
I'm slightly concerned this test for just localhost isn't robust enough and we should be using a better mechanism that will support loopback in general (i.e. not just those specific addresses, and also IPv6).

I know there's InetAddress#isAnyLocalAddress and InetAddress#isLoopbackAddress. Thoughts?

jhauray replied:
my apologies for the style. I will correct the spaces.

About the test for localhost, you're right. We must consider the local loopback and local addresses, whatever the expression. I will test `netAddress#isAnyLocalAddress` and `InetAddress#isLoopbackAddress`.

jhauray replied:
`InetAddress#isLoopbackAddress` is detected well loopback by ip or hostname.

But `netAddress#isAnyLocalAddress` is not detecting local machine addresses. It report if ip is wildcard, as 0.0.0.0.

So, to check if artifact url hostname is local address, I build a list of local ip addresses, for comparison.

msavy commented on gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java:
```diff
@@ -72,30 +93,73 @@ public VertxPluginRegistry(Vertx vertx, VertxEngineConfig vxEngineConfig, Map<St
+                    response.bodyHandler((Handler<Buffer>) buffer -> {
+                        try {
+                            // If status code is bad, do not handle the buffer.
+                            if(response.statusCode() != 200 ){
```
What non-200 code are you receiving in the main response that still causes a body to be received here? i.e. why the special handling?

Is this because you need to handle 20x rather than just 200?

jhauray replied:
During tests, I received 404 errors from Maven Central by querying unexisting artifact, example : https://repo1.maven.org/maven2/io/apiman/test/testPlugin/1.0.0.Final/testPlugin-1.0.0.Final.war

In this case, http status code is 404, and body content is (content-length : 554) :
```html
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
```

With previous code, whatever status code, Vert.X gateway try anyway to extract plugin from the body. Il was reporting a wrong error "error in opening zip file".

By handling 200 code, we could be sure that the body really contain à valid plugin artifact.

Notate that the same control already exist in DefaultPluginRegistry :
https://github.com/apiman/apiman/blob/ddaf009dcb06fc6924616bfa63f725d090f62792/gateway/engine/core/src/main/java/io/apiman/gateway/engine/impl/DefaultPluginRegistry.java#L376-L387


DefaultPluginRegistry.java

msavy replied:
Okay, I see. We should probably handle non-20x as an error condition.

EricWittmann commented on distro/vertx/src/main/resources/overlay/configs/conf-3scale-immutable-es.json:
```diff
@@ -72,7 +72,10 @@
       "class": "io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry",
-      "config": {}
+      "config": {
+        //"pluginRepositories":["https://url.to.maven.repo"],
```
Does this work?  Comments in JSON?

jhauray replied:
Normally, in pure JSON no. But in Vert.X, parser seems to be able to handle C-style comments.

It is described in Vert.X documentation : https://apiman.gitbooks.io/apiman-installation-guide/installation-guide/vertx/config-guide.html

EricWittmann replied:
Cool!

EricWittmann commented:
@jhauray and @msavy - are we ready to go on this?

jhauray commented:
@EricWittmann,

I let @msavy resolve the conversations if my feedback suits him.

About my last commit: It follows tests which showed that, following the errors returned by Maven repos (example Nexus), all the repos configured are not always tested. `result.isSuccess()` was not correctly tested.

I promise not to change anything anymore :-)

jhauray commented:
Oups,

I changed one last things : I added a unit test corresponding to my last fix.

Sorry for this new add @msavy... But I promise, it's finished.

jhauray commented:
> I think this looks good now.
> 
> The only thing I'm not sure about is the behaviour on Windows.
> 
> Have any of you tried this on Win? I don't have any Windows boxes to try it on.

Yes, I tried on Windows. My developement pc is under Windows. And I first run unit test on this computer before pushing on corporate CI/CD and github/travis.

I fixed a problem this morning. Our CI/CD settings changed. Unit test failed because lacking https.nonProxyHosts and http.nonProxyHosts proxy parameters. I just adding support for these parameters.

jhauray commented:
Issue #732 fixed.
